tesseract/tesseract_common/include/tesseract_common/interpolators.h

```diff
@@ -0,0 +1,169 @@
+                 const Eigen::Ref<Eigen::VectorXd>& derivatives,
+                 const Eigen::Ref<Eigen::VectorXi>& indices)
+    : x_min_(x_vec.minCoeff())
+    , x_max_(x_vec.maxCoeff())
+    ,
+    // Scale x values to [0:1] and curve fit with a 3rd order B-Spline.
+    spline_(Eigen::SplineFitting<Eigen::Spline<double, 1>>::InterpolateWithDerivatives(
+        y_vec.transpose(),
+        derivatives.transpose(),
+        indices,
+        std::min<unsigned>(static_cast<unsigned>(x_vec.rows()) - 1, 3),
+        scaledValues(x_vec)))
+  {
+    assert(derivatives.size() == indices.size());
+    assert(!(x_vec.array().isNaN().any()));
+    assert(!(x_vec.array().isInf().any()));
+    assert(!(y_vec.array().isNaN().any()));
+    assert(!(y_vec.array().isInf().any()));
+  }
+
+  /**
+   * @brief Constructor for interpolating with no restrictions on derivatives
+   * @param x_vec Independent variable. Probably time for most motion planning problems
+   * @param y_vec Dependent variable. Probably joint values for most motion planning problems.
+   */
+  SplineFunction(const Eigen::Ref<Eigen::VectorXd>& x_vec, const Eigen::Ref<Eigen::VectorXd>& y_vec)
+    : x_min_(x_vec.minCoeff())
+    , x_max_(x_vec.maxCoeff())
+    ,
+    // Scale x values to [0:1] and curve fit with a 3rd order B-Spline.
+    spline_(Eigen::SplineFitting<Eigen::Spline<double, 1>>::Interpolate(y_vec.transpose(),
+                                                                        std::min<long>(x_vec.rows() - 1, 3),
+                                                                        scaledValues(x_vec)))
+  {
+    assert(!(x_vec.array().isNaN().any()));
+    assert(!(x_vec.array().isInf().any()));
+    assert(!(y_vec.array().isNaN().any()));
+    assert(!(y_vec.array().isInf().any()));
+  }
+
+  /** @brief Returns the y value at point x in the spline */
+  double operator()(double x) const
+  {
+    // x values need to be scaled to [0:1] in extraction as well.
+    return spline_(scaledValue(x))(0);
+  }
+
+  /** @brief Returns a vector of interpolated y values for each x in the input vector */
+  Eigen::VectorXd operator()(const Eigen::Ref<Eigen::VectorXd>& x_vec) const
+  {
+    return x_vec.unaryExpr([this](double x) { return spline_(scaledValue(x))(0); });
+  }
+
+private:
+  /** @brief Scales value to [0:1] based on x_min and x_max */
+  double scaledValue(double x) const { return (x - x_min_) / (x_max_ - x_min_); }
+
+  /** @brief Scales vector such that each value is [0:1] based on x_min and x_max
+   *
+   * It is a requirement for Interpolate that x be on the interval [0:1] */
+  Eigen::RowVectorXd scaledValues(const Eigen::Ref<Eigen::VectorXd>& x_vec) const
+  {
+    return x_vec.unaryExpr([this](double x) { return scaledValue(x); }).transpose();
+  }
+
+  /** @brief Minimum value in x_vec. Used to scale inputs to [0:1] */
+  double x_min_;
+  /** @brief Maximum value in x_vec. Used to scale inputs to [0:1] */
+  double x_max_;
```

Comment:
The `x_vec` argument and `x_min`/`x_max` parameters (and associated scaling functions) seem a bit unnecessary in my opinion. Is there a compelling reason to use it? It seems like the user will almost always create a new vector of [0, 1, 2, ... n] like you did in the unit test. I think it would be more understandable if `x_vec` wasn't required in the constructor, and the user could only query the `()` operator with a value on [0:1]

Reply:
If you were to use this with Trajopt with time parameterization turned on (which coincidentally the interpolate function below is not set up for), you'd get joints with timesteps that are not evenly spaced. Then if you had a trajectory with a bunch of points clustered in time, you'd like to fit a spline to allow of those and then uniformly sample the whole thing in time.

Reply:
I would propose having the user provide a number on [0,1] when calling the `()` operator, since that is the common convention for splines anyway. Inferring max and min values from this seemingly unrelated `x_vec` input and silently scaling the argument to the `()` operator seems a bit confusing to me. Regarding the time parameterization example, I would assume that the calling code would convert the timestep to be on [0,1] before sampling the spline